src/workerd/server/container-client.c++

```diff
@@ -0,0 +1,406 @@
+    case 25:
+      return "SIGXFSZ"_kj;  // File size limit exceeded
+    case 26:
+      return "SIGVTALRM"_kj;  // Virtual alarm clock
+    case 27:
+      return "SIGPROF"_kj;  // Profiling alarm clock
+    case 28:
+      return "SIGWINCH"_kj;  // Window size change
+    case 29:
+      return "SIGIO"_kj;  // I/O now possible
+    case 30:
+      return "SIGPWR"_kj;  // Power failure restart (Linux)
+    case 31:
+      return "SIGSYS"_kj;  // Bad system call
+    default:
+      return "SIGKILL"_kj;
+  }
+}
+
+ContainerClient::ContainerClient(capnp::ByteStreamFactory& byteStreamFactory,
+    kj::Timer& timer,
+    kj::Network& network,
+    kj::String dockerPath,
+    kj::String containerName,
+    kj::String imageName)
+    : byteStreamFactory(byteStreamFactory),
+      timer(timer),
+      network(network),
+      dockerPath(kj::mv(dockerPath)),
+      containerName(kj::mv(containerName)),
+      imageName(kj::mv(imageName)) {}
+
+// Docker-specific Port implementation that implements rpc::Container::Port::Server
+class ContainerClient::DockerPort final: public rpc::Container::Port::Server {
+ public:
+  DockerPort(ContainerClient& containerClient, kj::String containerHost, uint16_t containerPort)
+      : containerClient(containerClient),
+        containerHost(kj::mv(containerHost)),
+        containerPort(containerPort) {}
+
+  kj::Promise<void> connect(ConnectContext context) override {
+    kj::HttpHeaderTable headerTable;
+    kj::HttpHeaders headers(headerTable);
+
+    // Port mappings might have outdated mappings, we can't know if a connect request
+    // fails because the app hasn't finished starting up or because the mapping is outdated.
+    // To be safe we should inspect the container to get up to date mappings.
+    auto portMappings = kj::get<1>(co_await containerClient.inspectContainer());
+    auto maybeMappedPort = portMappings.find(containerPort);
+    if (maybeMappedPort == kj::none) {
+      throw JSG_KJ_EXCEPTION(DISCONNECTED, Error,
+          "connect(): Connection refused: container port not found. Make sure you exposed the port in your container definition.");
+    }
+    auto mappedPort = KJ_ASSERT_NONNULL(maybeMappedPort);
+
+    auto address =
+        co_await containerClient.network.parseAddress(kj::str(containerHost, ":", mappedPort));
+    auto connection = co_await address->connect();
+
+    auto upPipe = kj::newOneWayPipe();
+    auto upEnd = kj::mv(upPipe.in);
+    auto results = context.getResults();
+    results.setUp(containerClient.byteStreamFactory.kjToCapnp(kj::mv(upPipe.out)));
+    auto downEnd = containerClient.byteStreamFactory.capnpToKj(context.getParams().getDown());
+    pumpTask =
+        kj::joinPromisesFailFast(kj::arr(upEnd->pumpTo(*connection), connection->pumpTo(*downEnd)))
+            .ignoreResult()
+            .attach(kj::mv(upEnd), kj::mv(connection), kj::mv(downEnd));
+    co_return;
+  }
+
+ private:
+  ContainerClient& containerClient;
```

Comment:
Given that this is a bare reference, a comment here indicating what owns the `ContainerClient` and keeps it alive would be helpful for others coming into this code later.

Reply:
Definitely. Nice catch. Thanks @jasnell.